File: 01.02.判定是否互为字符重排/01.02.cpp
// 解法1：两个map
class Solution {
public:
    bool CheckPermutation(string s1, string s2) {
        if (s1.size() != s2.size()) {
            return false;
        }

        unordered_map<char, int> map1;
        unordered_map<char, int> map2;
        
        for (char c : s1) {
            if (map1[c] > 0) {
                map1[c]++;
            } else {
                map1[c] = 1;
            }
        }

        for (char c : s2) {
            if (map2[c] > 0) {
                map2[c]++;
            } else {
                map2[c] = 1;
            }
        }

        return map1 == map2;
    }
};

// 解法2：字符串排序
class Solution {
public:
    bool CheckPermutation(string s1, string s2) {
        if (s1.size() != s2.size()) {
            return false;
        }
        
        sort(s1.begin(), s1.end());
        sort(s2.begin(), s2.end());
        return s1 == s2;
    }
};

File: 01.02.判定是否互为字符重排/01.02.go
// 解法1：两个map
func CheckPermutation(s1 string, s2 string) bool {
    if len(s1) != len(s2) {
        return false
    }

    m1, m2 := make(map[rune]int), make(map[rune]int)
    for _, r := range []rune(s1) {
        if m1[r] > 0 {
            m1[r]++
        } else {
            m1[r] = 1
        }
    }
    for _, r := range []rune(s2) {
        if m2[r] > 0 {
            m2[r]++
        } else {
            m2[r] = 1
        }
    }

    return reflect.DeepEqual(m1, m2)
}

// 解法2：字符串排序
func CheckPermutation(s1 string, s2 string) bool {
    if len(s1) != len(s2) {
        return false
    }

    r1, r2 := []rune(s1), []rune(s2)
    sort.Slice(r1, func(i int, j int) bool { return r1[i] < r1[j] })
    sort.Slice(r2, func(i int, j int) bool { return r2[i] < r2[j] })

    return reflect.DeepEqual(r1, r2)
}